src/hal/lv_hal_tick.c

```diff
@@ -29,8 +30,13 @@
  *  STATIC VARIABLES
  **********************/
 #if !LV_TICK_CUSTOM
-    static uint32_t sys_time = 0;
-    static volatile uint8_t tick_irq_flag;
+    #if LV_USE_ATOMICS == 1
+        static _Atomic(uint32_t) sys_time = 0;
```

Comment:
I wasn't atomic previously. Was it changed intentionally?

Reply:
Yes, because `lv_tick_inc`, which should be thread-safe, has a read-write access to the global `sys_time` variable. 
Maybe it should be declared volatile as well for the non-atomic fallback?

Reply:
There is a trick to make `lv_tick_inc()` and `sys_time` "thread-safe". See [here](https://github.com/lvgl/lvgl/blob/d979967d38990ced3c4a4ac6d0e967a46909ba07/src/hal/lv_hal_tick.c#L64-L73). I think this trick is required because if atomic is not supported on a 16 bit system, an `uint32_t` variable acn be read/written only in 2 steps. So the `lv_tick_inc()` interrupt can come in the wrong time.

It's not the case with the flush related variables as they are `int` and we can assume that they are written by one instruction.

What do you think?

Reply:
Yes, this might reduce the risk on 16-bit systems for a corrupt read. In case of  `stdatomic` support, we wouldn't need that work-around at all, so in that case, if `LV_USE_ATOMICS == 1` the function could just return `sys_time` in any case, even for 16-bit systems. 

But here are two misunderstandings. 1 assembly instruction does not necessarily mean atomic, it is important though that it "appears atomic" aka. the CPU is in the right mode. E.g. ARM is adding **additional** instructions in case of atomic operations: [[ldrex, strex]](https://developer.arm.com/documentation/dht0008/a/arm-synchronization-primitives/exclusive-accesses/ldrex-and-strex?lang=en)

And still `volatile` gives no synchronization guarantees between multiple threads, so it is technically not enough. E.g. on sophisticated systems with multiple CPUs `sys_time` or `tick_irq_flag` could be cached by the core cache, but the volatile access does not lead into synchronizing these caches. 

So it might work on systems with one CPU no caching and when `lv_tick_inc()` is called in an ISR. But with atomics, is safe to call in any circumstances. It would even be safe to call `lv_tick_inc()` from multiple threads or when `lv_tick_get()` could pre-empt `lv_tick_inc()` (assuming I'll remove the `tick_irq_flag` workaround in case of `LV_USE_ATOMICS == 1`).

Reply:
Okay, so after some reading I'm not to sure, if  `_Atomic(uint32_t)` compiles on every 16-bit platform, as the CPU might not support atomics `> uintmax_t`. But this error might be only reported by the final linking script, when it tries to find a polyphill lock for those atomic? (I am really not sure)> 

So the safe bet is to revert the sys_time -> atomic change, and keep the `tick_irq_flag` while loop, but still make that flag atomic.

Reply:
> So the safe bet is to revert the sys_time -> atomic change, and keep the tick_irq_flag while loop, but still make that flag atomic

I think it's a good compromise. It will work on simple embedded system (1 CPU) and on larger system atomic will be supported.

Reply:
Okay, I dug into this a little further: 

So, the `LV_USE_ATOMICS` check ensures, that `stdatomic` and the `_Atomic` keyword exists. But this is only a check if the compiler supports atomic. 

Now it comes down to the architecture if it does support atomic instructions as well. E.g. ARM with the [[ldrex, strex]](https://developer.arm.com/documentation/dht0008/a/arm-synchronization-primitives/exclusive-accesses/ldrex-and-strex?lang=en) and memory barrier support. But even for these architectures it comes down to the size of the type to be atomic, which mostly depends on the address width of the architecture as well. 

This property can be checked by the [`atomic_is_lock_free`](https://en.cppreference.com/w/c/atomic/atomic_is_lock_free) check.

So for example for [`avr` (godbolt)](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:7,endLineNumber:12,positionColumn:7,positionLineNumber:12,selectionStartColumn:7,selectionStartLineNumber:12,startColumn:7,startLineNumber:12),source:'%23include+%3Cstdint.h%3E%0A%23include+%3Cstdatomic.h%3E%0A%23include+%3Cstdio.h%3E%0A%0A%23if+HAVE_ATOMIC_FETCH_OP%0A%23error+bla%0A%23endif%0A%0A_Atomic(uint16_t)+num%3B%0A%0A/*+Type+your+code+here,+or+load+an+example.+*/%0Auint16_t+times2(void)+%7B%0A++++return+num+%2B%3D+num%3B%0A%7D%0A%0Aint+main()+%7B%0A++++printf(%22%25d%22,+atomic_is_lock_free(%26num))%3B%0A++++printf(%22%25lu%22,+times2())%3B%0A%7D%0A'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:34.176891489942705,l:'4',m:100,n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cavrg1220,deviceViewOpen:'1',filters:(b:'0',binary:'0',binaryObject:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'0',intel:'1',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-latomic+-O3',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+AVR+gcc+12.2.0+(Editor+%231)',t:'0')),header:(),k:35.922218641626074,l:'4',m:100,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+12.2',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+AVR+gcc+12.2.0+(Compiler+%231)',t:'0')),k:29.900889868431236,l:'4',m:100,n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) there is no builtin atomic support so functions like `__atomic_fetch_add_2` are always called, independent on the size of the atomic variable. 

For [ARM (godbolt)](https://godbolt.org/#g:!((g:!((g:!((h:compiler,i:(compiler:cclang1500,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'--target%3Dthumbv7em-none-eabihf',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,wantOptInfo:'1'),l:'5',n:'0',o:'+x86-64+clang+15.0.0+(Editor+%231)',t:'0')),header:(),k:50,l:'4',m:100,n:'0',o:'',s:0,t:'0'),(g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:24,endLineNumber:14,positionColumn:24,positionLineNumber:14,selectionStartColumn:24,selectionStartLineNumber:14,startColumn:24,startLineNumber:14),source:'%23include+%3Cstdint.h%3E%0A%23include+%3Cstdatomic.h%3E%0A//+%23include+%3Cstdio.h%3E%0A%0A_Atomic(uint32_t)+num%3B%0A_Atomic(uint64_t)+num2%3B%0A%0A/*+Type+your+code+here,+or+load+an+example.+*/%0Auint32_t+times2(void)+%7B%0A++++return+num+%2B%3D+num%3B%0A%7D%0A%0Auint32_t+times2_64(void)+%7B%0A++++return+num2+%2B%3D+num2%3B%0A%7D%0A%0A'),l:'5',n:'0',o:'C+source+%231',t:'0')),header:(),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) however, it depends on the size of the variable. 

Anyway, in cases of calling the builtin `__atomic_*` functions it is not guaranteed to be lockless anymore, and `-latomic` aka. libatomic must be linked, so that the program can compile. AFAIK, avr provides no `libatomic` by default, and it must be user provided. Meaning, we can not assume, that this will compile for every architecture.

With that knowledge, I worry we can't just enable using atomics by default, as this might break user code, which worked upon that point, because they did not provide any `libatomic` or just did not add the `-latomic` linker flag. 

Ideally, such support should be tested by built tools like cmake before automatically enabling it, but this heavily on the user use-case. I'd propose to add the `LV_USE_ATOMICS` options back to `lv_conf.h`.
We can still default it to `1` aka enabled (if a breaking change is ok), but the user at least has a chance to disable it, if they did not implement libatomic.